src/function.jl

```diff
@@ -0,0 +1,19 @@
+abstract type AbstractOptimizationFunction end
+
+struct OptimizationFunction{F,G,H,K} <: AbstractOptimizationFunction
+    f::F
+    grad::G
+    hess::H
+    kwargs::K
+end
+
+function OptimizationFunction(f; grad=nothing,hess=nothing, p=DiffEqBase.NullParameters(),kwargs...)
+    if p isa DiffEqBase.NullParameters
+        g! = (res,x) -> ForwardDiff.gradient!(res, f, x)
```

Comment:
this branch isn't needed

Reply:
How would this be handled? We can expect the user to always pass in the `x -> f(x,p)`?

Reply:
always expect `f(x,p)`